### PR TITLE
torch.compile + selective activation checkpointing

### PR DESCRIPTION
**Pull request by yf225**
Pull Request resolved: https://github.com/pytorch/pytorch/pull/105489

NOTE: this PR is tagged "not user facing", because it's not ready to be announced externally yet.

This PR implements torch.compile + selective activation checkpoint (SAC) integration, by using `TagActivationCheckpoint` (same backend as torch.compile + full activation checkpoint integration).

TorchDispatchMode based implementation cannot support including inplace ops in the checkpointed region at the moment (the reason for this needs investigation), and there is also no way to ban them (because TorchDispatchMode now only sees "after-functionalization" ops, so can't detect if an op is in-place). Hence we hide torch.compile + SAC behind a flag (`torch._dynamo.config._experimental_support_context_fn_in_torch_utils_checkpoint`) and will only use it internally for cases that are known to not have in-place ops. This state won't last too long, because in-place op will at least be able to be detected after Brian's mode reordering and related functionalization changes.
So next steps after this PR:
1. Wait for Brian's mode reordering and related functionalization changes to land, and then try to enable the "inplace ops" unit test for torch.compile + selective activation checkpoint (if it doesn't work, investigate why).
2. Unify selective- and full-checkpoint under TorchDispatchMode based implementation.

Differential Revision: D47497145

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @anijain2305 @ipiszy

### CODE REVIEW

File: torch/_higher_order_ops/wrap.py
Comment (anijain2305) [inline torch/_higher_order_ops/wrap.py:494]: IMO, having a switch to explicitly go one way or the other is a better approach. This is with the assumption that SAC is a niche usecase.

AOT Autograd traced checkpoint requires more work around CSE etc. As that part of the stack is not hardened. I think its better if we have well tested path as default. And when that path does not work out, we emit an actionable error message to switch to the less tested path.

**Comment by Chillee** [comment]
I don't think we actually need to ban in-place ops right? Unless they modify one of the inputs?

File: torch/_higher_order_ops/wrap.py
Comment (yf225) [reply torch/_higher_order_ops/wrap.py:494]: CSE seems to not cause problem with selective checkpointing right now actually, but yes I can make this into a flag.

File: test/dynamo/test_activation_checkpointing.py
Comment (yf225) [inline test/dynamo/test_activation_checkpointing.py:625]: @Chillee I compared this test case with `test_tags_selective_checkpoint_outplace_op` above (`torch.selu_` vs. `torch.selu`), and it seems that as long as we have in-place op applied to inputs or activations in the graph, the gradient comparison check would fail.

Comment (Chillee) [reply test/dynamo/test_activation_checkpointing.py:625]: Hmm... I need to think a bit more about this. I would ... expect that functionalization would apply underneath selective checkpointing, and so it would still work correctly.

File: torch/_higher_order_ops/wrap.py
Comment (albanD) [inline torch/_higher_order_ops/wrap.py:33]: Can we find a way not to hardcode these?

File: torch/_higher_order_ops/wrap.py
Comment (albanD) [inline torch/_higher_order_ops/wrap.py:68]: `if name[-1] == "_" or (name.starts_with("__i") and name.ends_with("__")):`

File: torch/_dynamo/variables/higher_order_ops.py
Comment (Chillee) [inline torch/_dynamo/variables/higher_order_ops.py:895]: So this works with arbitrary context_fns?

Comment (yf225) [reply torch/_dynamo/variables/higher_order_ops.py:895]: Only guarantees to work with TorchDispatch based context_fns actually. Maybe we can assert that user must use TorchDispatch functions?

File: test/dynamo/test_activation_checkpointing.py
Comment (yf225) [inline test/dynamo/test_activation_checkpointing.py:66]: Note to self: updated to use detach similar to https://fburl.com/code/oaxjq5n6

File: test/dynamo/test_activation_checkpointing.py
Comment (yf225) [inline test/dynamo/test_activation_checkpointing.py:462]: Note to self: updated to use detach similar to https://fburl.com/code/oaxjq5n6

File: torch/_higher_order_ops/wrap.py
Comment (anijain2305) [reply torch/_higher_order_ops/wrap.py:494]: Have you checked the memory numbers after torch.compile? CSE will silently undo AC, and will increase memory footprint.

File: test/dynamo/test_activation_checkpointing.py
Comment (bdhirsh) [reply test/dynamo/test_activation_checkpointing.py:625]: Hmm. This seems pretty similar to the "what should run first, torch_dispatch vs. functionalization" ordering issue I'm running into with subclass support. I don't think this will work today, since the "default" thing that happens is that functionalization has a higher priority in the dispatcher than __toch_dispatch__.

I updated the design doc after running into some issues with DTensor - I think functionalization will need to run below torch_dispatch (updated [here](https://docs.google.com/document/d/1COQ291nOZvtFnzGTQMJqoYZ3sttEYFw_7HbfSyL8gcA/edit#bookmark=id.4tvs5sten7fb)). Hopefully that should fix this problem too - I can give a shout once I have a working PR to test with if that sounds good to you all.

Comment (bdhirsh) [reply test/dynamo/test_activation_checkpointing.py:625]: But (once the functionalization ordering issue is sorted out), agreed with Horace's comment - instead of manually checking for inplace ops, would it make sense to use the existing `speculate_graph` infra that Voz/Richard wrote, that just tries to ban subgraphs containing input mutations?

File: torch/_higher_order_ops/wrap.py
Comment (bdhirsh) [reply torch/_higher_order_ops/wrap.py:33]: As an alternative (just as an idea, not sure that this is the best approach):

Unfortunately, we don't have a systematic way to get all random ops from the `torch.*` module. But if instead you could get at the ATen ops that will be captured in our checkpointed function, then you can inspect the `tags` on each the `OpOverload` for a `nondeterminism` tag (indicated randomness, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/tags.yaml#L25))

If we have already used dynamo to confirm that the function is traceable, then one option would be to `make_fx()` it and get out a graph of the aten ops.

File: test/dynamo/test_activation_checkpointing.py
Comment (yf225) [inline test/dynamo/test_activation_checkpointing.py:603]: Note to self: check that we are recomputing selu

File: torch/_higher_order_ops/wrap.py
Comment (yf225) [reply torch/_higher_order_ops/wrap.py:494]: Note to self: check memory usage of SAC under torch.compile

Also found that `min_cut_rematerialization_partition` (which has CSE) is not used when SAC is applied, and it might explain why we don't see CSE causing problem with SAC.

**Comment by yf225** [comment]
`fx_traceback.current_meta` is another way to track the recompute tag for user dispatch mode

File: test/dynamo/test_activation_checkpointing.py
Comment (yf225) [reply test/dynamo/test_activation_checkpointing.py:624]: yes - I can confirm that with Brian's mode reordering stack, I can use `func._schema.is_mutable` to check inplace ops.

Comment (bdhirsh) [reply test/dynamo/test_activation_checkpointing.py:624]: Hmm @yf225 do we want to ban **all** mutable ops for now? Or is that needed any time soon.

Functionalization should handle any mutable ops inside the checkpointed region. The main issue is just "what if a mutable op tries to mutate one of the **inputs** to the checkpointed region" - and in that case, higher order ops today are supposed to error in `speculate_subgraph` if they detect that there were any input mutations.

File: torch/_dynamo/variables/higher_order_ops.py
Comment (yf225) [reply torch/_dynamo/variables/higher_order_ops.py:1086]: Added check in torch/utils/checkpoint.py and corresponding unit test.

File: torch/_higher_order_ops/wrap.py
Comment (yf225) [reply torch/_higher_order_ops/wrap.py:150]: Update: I found a way to use `node.meta["recompute"]` for selective checkpointing case that's similar to full-checkpointing, details are in `CachedTorchDispatchMode._handle_compile_in_forward_ctx` method

File: torch/_higher_order_ops/wrap.py
Comment (yf225) [reply torch/_higher_order_ops/wrap.py]: Update: will use Brian's mode reordering stack, and check `.is_mutable` for inplace ops and `nondeterministic_seeded` tag for random ops

File: torch/utils/checkpoint.py
Comment (yf225) [inline torch/utils/checkpoint.py:1219]: Throw error (and add unit test) if user's own TDM doesn't handle compile properly

File: torch/utils/checkpoint.py
Comment (yf225) [inline torch/utils/checkpoint.py:1191]: After discussion: we should make Caching and Cached private APIs, and only expose `policy_fn` API to users.

File: torch/utils/checkpoint.py
Comment (yf225) [inline torch/utils/checkpoint.py:1176]: we should pass "forward" instead of "self" here

File: torch/utils/checkpoint.py
Comment (yf225) [inline torch/utils/checkpoint.py:1211]: we should pass "recompute" instead of "self" here

File: test/dynamo/test_activation_checkpointing.py
Comment (yf225) [inline test/dynamo/test_activation_checkpointing.py:475]:
```suggestion
                mm_count_key = f"{dispatch_mode}_mm_count"
```

**Comment by yf225** [comment]
@anijain2305 @Chillee I changed the approach so that selective checkpointing also uses the partitioner-based “duplicate late” approach, similar to full-checkpointing. Checked with Horace and he seems okay with it, and he said as long as we can later change back to “duplicate early” when things are ready. To support that vision, I only expose the `policy_fn` (func -> whether to checkpoint) as the public API to users, which is the API that will never change regardless of internal implementation.

File: test/dynamo/test_activation_checkpointing.py
Comment (anijain2305) [inline test/dynamo/test_activation_checkpointing.py:460]: You can annotate the test with `@torch._functorch.config.patch(cse=True)`

File: torch/_higher_order_ops/wrap.py
Comment (anijain2305) [inline torch/_higher_order_ops/wrap.py:45]: This is confusing. Why do we even need `recomputable` tags if we are relying on AOT Autograd to do the tracing.

Imo, we should keep WrapActivationCheckpoint simple - let AOT Autograd take care of everything. In fact, this TorchDispatch facility seems to sit well with TagActivationCheckpoint to me.

File: torch/_higher_order_ops/wrap.py
Comment (anijain2305) [inline torch/_higher_order_ops/wrap.py:133]: Yeah, this on face value seems very confusing. WrapActivationCheckpoint (and this is a bad name - my mistake) should not need any tagging. This should go in TagActivationCheckpoint.

File: torch/utils/checkpoint.py
Comment (anijain2305) [inline torch/utils/checkpoint.py:418]: Just interested - why dont we allow in place ops with SAC?

File: torch/utils/checkpoint.py
Comment (anijain2305) [inline torch/utils/checkpoint.py:1160]: Oh is that always the case? Can we never have random ops in selective checkpointing? Why is that the case?

The whole point of Duplicate Late is to handle rng ops. If no rng ops, Duplicate early + disabling cse might be good enough. So, this complexity of TorchDispatchMode and still using AOT Autograd for tracing seems unnecessary. I am still behind the idea of unifying but currently the stack seems little jarring.

File: torch/utils/checkpoint.py
Comment (yf225) [inline torch/utils/checkpoint.py:1156]: see if we can remove this inplace restriction

File: torch/utils/checkpoint.py
Comment (yf225) [inline torch/utils/checkpoint.py:1160]: see if we can remove this random op restriction

File: torch/_higher_order_ops/wrap.py
Comment (yf225) [inline torch/_higher_order_ops/wrap.py:142]: we should put SAC+compile logic into TagActivationCheckpoint (and leave the old WrapActivationCheckpoint mode as-is).

Also try to see if full-checkpoint can use the same mode-based `return checkpoint(Interpreter(gmod).run, *args, **kwargs)` logic (by having a special pair of modes for full case).

Comment (yf225) [inline torch/_higher_order_ops/wrap.py:142]: Also try to see if full-checkpoint can use the same mode-based `return checkpoint(Interpreter(gmod).run, *args, **kwargs)` logic (by having a special pair of modes for full case).

File: torch/utils/checkpoint.py
Comment (yf225) [reply torch/utils/checkpoint.py:1156]: we need Brian's stack to make this work. We should add a config to enable this only for experimental use cases.

File: test/dynamo/test_activation_checkpointing.py
Comment (yf225) [reply test/dynamo/test_activation_checkpointing.py]: Update: will wait for Brian's stack to land for enabling in-place op support

File: torch/utils/checkpoint.py
Comment (yf225) [reply torch/utils/checkpoint.py]: Currently running into gradient mismatch for the inplace test case we have, will wait for Brian's stack to land and then investigate this again

File: torch/utils/checkpoint.py
Comment (yf225) [reply torch/utils/checkpoint.py]: Removed restriction on random ops

File: torch/_higher_order_ops/wrap.py
Comment (anijain2305) [inline torch/_higher_order_ops/wrap.py:143]: Have we tried making this path also go through the `checkpoint`? If not too hard, we should make it happen. Maybe we can have dummy context_fn if there is no user passed context_fn

File: torch/utils/checkpoint.py
Comment (anijain2305) [inline torch/utils/checkpoint.py:1210]: It seems like we need this only for torch.compile (even though it can work w/o torch.compile as well). Maybe mention the reason of introducing this wrt torch.compile.

File: torch/_higher_order_ops/wrap.py
Comment (anijain2305) [inline torch/_higher_order_ops/wrap.py:73]: I think this is still partially true.

Dynamo still does not trace through saved tensor hooks.

AOT Autograd does trace through saved tensor hooks but has special logic with TorchDispatchMode to override the usual saved_tenso_hooks fn to tag the nodes.

The comment about RNG is still valid.

File: torch/_higher_order_ops/wrap.py
Comment (anijain2305) [inline torch/_higher_order_ops/wrap.py:126]: Maybe mention why do we restrict relu_ ops?

File: test/dynamo/test_activation_checkpointing.py
Comment (anijain2305) [inline test/dynamo/test_activation_checkpointing.py:419]: Do you need this?

File: test/dynamo/test_activation_checkpointing.py
Comment (anijain2305) [inline test/dynamo/test_activation_checkpointing.py:451]: nit - Just a comment that we would expect 6 if not for SAC.

File: test/dynamo/test_activation_checkpointing.py
Comment (anijain2305) [inline test/dynamo/test_activation_checkpointing.py:481]: Nice! This is powerful.

File: test/dynamo/test_activation_checkpointing.py
Comment (anijain2305) [inline test/dynamo/test_activation_checkpointing.py:514]: nit - I think its decided by the autograd dependencies, not really by the partitioner.

**Review by anijain2305 — COMMENTED** [review]
Overall LGTM. Great work! Some nits here and there.

The only real comment is about unifying the `checkpoint` tracing between no context_fn and yes context_fn. 

Please also get a review from @soulitzer @albanD on the checkpoint related files.

**Review by anijain2305 — APPROVED** [review]
Chatted offline. Just add a comment about unification dependent on the inplace support in TorchDispatch models.

**Pull request closed without merging**
